dask_cloudprovider/providers/gcp/instances.py

```diff
@@ -0,0 +1,326 @@
+import asyncio
+import uuid
+
+import dask
+from dask_cloudprovider.providers.generic.vmcluster import (
+    VMCluster,
+    VMInterface,
+    SchedulerMixin,
+    WorkerMixin,
+)
+
+from dask_cloudprovider.utils.socket import is_socket_open
+
+try:
+    import googleapiclient.discovery
+    from googleapiclient.errors import HttpError
+    from google.auth.exceptions import DefaultCredentialsError
+except ImportError as e:
+    msg = (
+        "Dask Cloud Provider GCP requirements are not installed.\n\n"
+        "Please either conda or pip install as follows:\n\n"
+        "  conda install -c conda-forge google-api-python-client  # either conda install\n"
+        '  python -m pip install "dask-cloudprovider[gcp]" --upgrade  # or python -m pip install'
+    )
+    raise ImportError(msg) from e
+
+
+class GCPInstance(VMInterface):
+    def __init__(
+        self,
+        cluster,
+        name,
+        config=None,
+        zone=None,
+        projectid=None,
+        machine_type=None,
+        filesystem_size=None,
+        source_image=None,
+        docker_image=None,
+        ngpus=None,
+        gpu_type=None,
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self.cluster = cluster
+        self.name = name
+        self.config = config
+        self.projectid = projectid or self.config.get("projectid")
```

Comment:
Can we autodetect a default project here? If this is not configured things fail.

```python-traceback
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-17-6a120a73072a> in async-def-wrapper()

/opt/conda-environments/rapids-stable/lib/python3.7/site-packages/distributed/deploy/spec.py in _()
    382         async def _():
    383             if self.status == Status.created:
--> 384                 await self._start()
    385             await self.scheduler
    386             await self._correct_state()

~/Projects/dask/dask-cloudprovider/dask_cloudprovider/providers/generic/vmcluster.py in _start(self)
    115             "Hang tight! ",
    116         ):
--> 117             await super()._start()
    118 
    119     def render_cloud_init(self, *args, **kwargs):

/opt/conda-environments/rapids-stable/lib/python3.7/site-packages/distributed/deploy/spec.py in _start(self)
    302 
    303         self.status = Status.starting
--> 304         self.scheduler = await self.scheduler
    305         self.scheduler_comm = rpc(
    306             getattr(self.scheduler, "external_address", None) or self.scheduler.address,

/opt/conda-environments/rapids-stable/lib/python3.7/site-packages/distributed/deploy/spec.py in _()
     69             async with self.lock:
     70                 if self.status == Status.created:
---> 71                     await self.start()
     72                     assert self.status == Status.running
     73             return self

~/Projects/dask/dask-cloudprovider/dask_cloudprovider/providers/gcp/instances.py in start(self)
    225     async def start(self):
    226         await super().start()
--> 227         await self.start_scheduler()
    228 
    229     async def start_scheduler(self):

~/Projects/dask/dask-cloudprovider/dask_cloudprovider/providers/gcp/instances.py in start_scheduler(self)
    237                            )
    238         self.cluster._log("Creating scheduler instance")
--> 239         self.internal_ip, self.external_ip = await self.create_vm()
    240         self.address = f"tcp://{self.external_ip}:8786"
    241         await self.wait_for_scheduler()

~/Projects/dask/dask-cloudprovider/dask_cloudprovider/providers/gcp/instances.py in create_vm(self)
    162             inst = (
    163                 self.cluster.compute.instances()
--> 164                 .insert(project=self.projectid, zone=self.zone, body=self.gcp_config)
    165                 .execute()
    166             )

~/.local/lib/python3.7/site-packages/googleapiclient/discovery.py in method(self, **kwargs)
    970                     _methodProperties(methodDesc, schema, "response")
    971                 ):
--> 972                     raise TypeError('Missing required parameter "%s"' % name)
    973 
    974         for name, regex in six.iteritems(parameters.pattern_params):

TypeError: Missing required parameter "project"
```

Reply:
It's not obvious to me what the default should be here.  When [listing project](https://cloud.google.com/sdk/gcloud/reference/projects/list) gcloud returns a list of all projects which the user has Owner, Editor, Viewer permissions of.  Perhaps it's better to error nicely ?